File: backend/app/engine/rrg.py
"""Relative Rotation Graph (RRG) multi-strike momentum engine.

Each option strike is treated as a "security" rotating against its index spot
("benchmark").  We use the JdK RS-Ratio / RS-Momentum construction in its
ratio-to-moving-average form:

1. ``RS_t = 100 × premium_t / spot_t`` — the raw relative-strength line.
2. ``RS-Ratio_t = 100 × RS_t / SMA(RS, window)`` — is relative strength above or
   below its own trend?
3. ``RS-Momentum_t = 100 × RS_t / RS_{t−k}`` — the rate of change of that same
   relative-strength line over the momentum lookback.

Both axes are therefore centred on 100 and scale-free, which is what makes the
quadrant test a simple comparison against 100.  Momentum is deliberately taken
on the RS line rather than on RS-Ratio: a rate of change of an *already
normalised* series saturates under a steady trend and parks genuinely trending
nodes back at the origin, and it is the raw ROC that turns first — which is what
produces the clockwise rotation the HUD draws.

Values are clipped so one illiquid print cannot fling a node off the plot, and a
bounded tail of prior points is retained per node for the HUD's momentum trace.
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field

from ..config import settings
from ..schemas import Quadrant, RrgPoint

#: Axis clip, in points either side of the 100 origin.
AXIS_CLIP = 12.0
#: Minimum observations before a node is considered statistically meaningful.
MIN_SAMPLES = 8


def classify_quadrant(rs_ratio: float, rs_momentum: float) -> Quadrant:
    """Map an (RS-Ratio, RS-Momentum) coordinate onto its RRG quadrant."""
    if rs_ratio >= 100.0:
        return Quadrant.LEADING if rs_momentum >= 100.0 else Quadrant.WEAKENING
    return Quadrant.IMPROVING if rs_momentum >= 100.0 else Quadrant.LAGGING


def _ratio_to_mean(series: deque[float]) -> float:
    """``100 × latest / mean(series)``, clipped to the readable axis range."""
    if not series:
        return 100.0
    mean = sum(series) / len(series)
    if abs(mean) <= 1e-12:
        return 100.0
    value = 100.0 * series[-1] / mean
    return max(100.0 - AXIS_CLIP, min(100.0 + AXIS_CLIP, value))


def _clip(value: float) -> float:
    return max(100.0 - AXIS_CLIP, min(100.0 + AXIS_CLIP, value))


@dataclass
class _NodeState:
    rs: deque[float]
    tail: deque[RrgPoint] = field(default_factory=deque)
    samples: int = 0
    last: RrgPoint | None = None


class RrgEngine:
    """Stateful, per-token RRG calculator.

    ``update`` is cheap (O(window)) and is called once per strike per stream
    frame, so the whole 4-quadrant matrix refreshes in a single pass.
    """

    def __init__(
        self,
        window: int | None = None,
        momentum_lookback: int | None = None,
        tail_length: int | None = None,
    ) -> None:
        self.window = window or settings.rrg_window
        self.momentum_lookback = momentum_lookback or settings.rrg_momentum_lookback
        self.tail_length = tail_length or settings.rrg_tail_length
        self._nodes: dict[str, _NodeState] = {}

    # ------------------------------------------------------------------ #
    def _state(self, token: str) -> _NodeState:
        state = self._nodes.get(token)
        if state is None:
            state = _NodeState(
                rs=deque(maxlen=max(self.window, self.momentum_lookback + 1)),
                tail=deque(maxlen=self.tail_length),
            )
            self._nodes[token] = state
        return state

    def update(self, token: str, price: float, benchmark: float) -> RrgPoint:
        """Feed one observation and return the node's current RRG coordinate."""
        state = self._state(token)
        if price <= 0 or benchmark <= 0:
            return state.last or RrgPoint(rs_ratio=100.0, rs_momentum=100.0)

        state.samples += 1
        state.rs.append(100.0 * price / benchmark)

        rs_ratio = _ratio_to_mean(state.rs)

        if len(state.rs) > self.momentum_lookback:
            past = state.rs[-1 - self.momentum_lookback]
            rs_momentum = (
                _clip(100.0 * state.rs[-1] / past) if abs(past) > 1e-12 else 100.0
            )
        else:
            rs_momentum = 100.0

        # Until we have a meaningful sample, pull the node toward the origin so
        # the HUD does not present noise as conviction.
        if state.samples < MIN_SAMPLES:
            damp = state.samples / MIN_SAMPLES
            rs_ratio = 100.0 + (rs_ratio - 100.0) * damp
            rs_momentum = 100.0 + (rs_momentum - 100.0) * damp

        point = RrgPoint(rs_ratio=round(rs_ratio, 4), rs_momentum=round(rs_momentum, 4))
        state.tail.append(point)
        state.last = point
        return point

    def point(self, token: str) -> RrgPoint | None:
        state = self._nodes.get(token)
        return state.last if state else None

    def quadrant(self, token: str) -> Quadrant | None:
        p = self.point(token)
        return classify_quadrant(p.rs_ratio, p.rs_momentum) if p else None

    def tail(self, token: str) -> list[RrgPoint]:
        state = self._nodes.get(token)
        return list(state.tail) if state else []

    def matured(self, token: str) -> bool:
        state = self._nodes.get(token)
        return bool(state and state.samples >= MIN_SAMPLES)

    def reset(self, token: str | None = None) -> None:
        if token is None:
            self._nodes.clear()
        else:
            self._nodes.pop(token, None)

File: web/lib/engine/rrg.ts
import type { Quadrant, RrgPoint } from "@/lib/types";

/**
 * Relative Rotation Graph engine — TypeScript port of `backend/app/engine/rrg.py`.
 *
 * Each option strike rotates against its index spot:
 *   RS          = 100 × premium / spot
 *   RS-Ratio    = 100 × RS / mean(RS, window)      — trend position
 *   RS-Momentum = 100 × RS / RS[t−k]               — rate of change of that line
 *
 * Momentum is taken on the RS line rather than on RS-Ratio deliberately: the
 * rate of change of an already-normalised series saturates under a steady trend
 * and parks genuinely trending nodes back at the origin. The raw ROC also turns
 * first, which is what produces the clockwise rotation the HUD draws.
 */

export const AXIS_CLIP = 12;
export const MIN_SAMPLES = 8;

export function classifyQuadrant(
  rsRatio: number,
  rsMomentum: number,
): Quadrant {
  if (rsRatio >= 100) return rsMomentum >= 100 ? "LEADING" : "WEAKENING";
  return rsMomentum >= 100 ? "IMPROVING" : "LAGGING";
}

/**
 * Soft-bounded rather than hard-clamped: a hard `Math.max`/`Math.min` pins
 * every node beyond the edge to the exact same coordinate, so a strongly
 * leading node and a wildly leading node land on top of each other and read
 * as identical. `tanh` approaches the same ±`AXIS_CLIP` bound asymptotically
 * without ever flattening it — nodes stay visually ordered by how extreme
 * they actually are, all the way to the edge.
 */
const clip = (v: number) => 100 + AXIS_CLIP * Math.tanh((v - 100) / AXIS_CLIP);

function ratioToMean(series: number[]): number {
  if (series.length === 0) return 100;
  const mean = series.reduce((a, b) => a + b, 0) / series.length;
  if (Math.abs(mean) <= 1e-12) return 100;
  return clip((100 * series[series.length - 1]) / mean);
}

interface NodeState {
  rs: number[];
  tail: RrgPoint[];
  samples: number;
  /** The last price a `samples` increment was actually earned by. */
  lastPrice: number | null;
  last: RrgPoint | null;
}

export class RrgEngine {
  private nodes = new Map<string, NodeState>();

  constructor(
    private window = 40,
    private momentumLookback = 5,
    private tailLength = 12,
    /** Genuine-price-change threshold before a node matures. Per-instance so a thinly-traded underlying can mature sooner than the shared default. */
    private minSamples = MIN_SAMPLES,
  ) {}

  private state(token: string): NodeState {
    let s = this.nodes.get(token);
    if (!s) {
      s = { rs: [], tail: [], samples: 0, lastPrice: null, last: null };
      this.nodes.set(token, s);
    }
    return s;
  }

  update(token: string, price: number, benchmark: number): RrgPoint {
    const s = this.state(token);
    if (price <= 0 || benchmark <= 0) {
      return s.last ?? { rs_ratio: 100, rs_momentum: 100 };
    }

    /**
     * The caller advances every node once a second whenever *anything* in
     * the whole tick universe printed, not when this particular contract
     * did — an untraded strike's `price` is simply carried forward, so most
     * calls here repeat the same value rather than deliver new information.
     * Counting those as samples let a leg that hasn't traded in a minute
     * still reach `MIN_SAMPLES` on the clock alone, and get plotted with the
     * same confidence as one that is actually printing. Only a genuine price
     * change earns a sample; the window below still advances every call, so
     * the ratio's mean stays a real rolling average rather than one padded
     * with copies of a stale print.
     */
    if (s.lastPrice === null || price !== s.lastPrice) {
      s.samples += 1;
      s.lastPrice = price;
    }
    s.rs.push((100 * price) / benchmark);
    const cap = Math.max(this.window, this.momentumLookback + 1);
    if (s.rs.length > cap) s.rs.shift();

    let rsRatio = ratioToMean(s.rs);

    let rsMomentum = 100;
    if (s.rs.length > this.momentumLookback) {
      const past = s.rs[s.rs.length - 1 - this.momentumLookback];
      if (Math.abs(past) > 1e-12) {
        rsMomentum = clip((100 * s.rs[s.rs.length - 1]) / past);
      }
    }

    // Until we have a meaningful sample, pull the node toward the origin so the
    // HUD does not present noise as conviction.
    if (s.samples < this.minSamples) {
      const damp = s.samples / this.minSamples;
      rsRatio = 100 + (rsRatio - 100) * damp;
      rsMomentum = 100 + (rsMomentum - 100) * damp;
    }

    const point: RrgPoint = {
      rs_ratio: Number(rsRatio.toFixed(4)),
      rs_momentum: Number(rsMomentum.toFixed(4)),
    };
    s.tail.push(point);
    if (s.tail.length > this.tailLength) s.tail.shift();
    s.last = point;
    return point;
  }

  point(token: string): RrgPoint | null {
    return this.nodes.get(token)?.last ?? null;
  }

  quadrant(token: string): Quadrant | null {
    const p = this.point(token);
    return p ? classifyQuadrant(p.rs_ratio, p.rs_momentum) : null;
  }

  tail(token: string): RrgPoint[] {
    return this.nodes.get(token)?.tail.slice() ?? [];
  }

  matured(token: string): boolean {
    const s = this.nodes.get(token);
    return !!s && s.samples >= this.minSamples;
  }

  reset(token?: string): void {
    if (token === undefined) this.nodes.clear();
    else this.nodes.delete(token);
  }
}
